Firestore/Source/API/FIRFieldPath.mm

```diff
@@ -16,34 +16,55 @@
 
 #import "Firestore/Source/API/FIRFieldPath+Internal.h"
 
-#import "Firestore/Source/Model/FSTPath.h"
+#include <functional>
+#include <string>
+#include <utility>
+#include <vector>
+
 #import "Firestore/Source/Util/FSTUsageValidation.h"
 
+#include "Firestore/core/src/firebase/firestore/model/field_path.h"
+#include "Firestore/core/src/firebase/firestore/util/string_apple.h"
+
+namespace util = firebase::firestore::util;
+using firebase::firestore::model::FieldPath;
+
 NS_ASSUME_NONNULL_BEGIN
 
+@interface FIRFieldPath () {
+  /** Internal field path representation */
+  firebase::firestore::model::FieldPath _internalValue;
+}
+
+@end
+
 @implementation FIRFieldPath
 
 - (instancetype)initWithFields:(NSArray<NSString *> *)fieldNames {
   if (fieldNames.count == 0) {
     FSTThrowInvalidArgument(@"Invalid field path. Provided names must not be empty.");
   }
 
+  std::vector<std::string> field_names{};
+  field_names.reserve(fieldNames.count);
   for (int i = 0; i < fieldNames.count; ++i) {
     if (fieldNames[i].length == 0) {
       FSTThrowInvalidArgument(@"Invalid field name at index %d. Field names must not be empty.", i);
     }
+    field_names.emplace_back([fieldNames[i] UTF8String],
+                             [fieldNames[i] lengthOfBytesUsingEncoding:NSUTF8StringEncoding]);
   }
 
-  return [self initPrivate:[FSTFieldPath pathWithSegments:fieldNames]];
+  return [self initPrivate:FieldPath(std::move(field_names))];
```

Comment:
comment: I think copy elision makes std::move redundant here. But I have no objection to keeping this either. (Also on line 67 below. Similarly no objection.)

Reply:
ack